src/Share/NamespaceDiffs.hs

```diff
@@ -623,22 +809,37 @@ computeThreeWayNamespaceDiff codebaseEnvs2 branchHashIds3 nameLookupReceipts3 =
           (TermReferenceId, (Term Symbol Ann, Type Symbol Ann))
           (TypeReferenceId, Decl Symbol Ann)
       ) <- do
-    let hydrateTerm ::
-          Codebase.CodebaseEnv ->
-          TermReferenceId ->
-          PG.Transaction e (TermReferenceId, (Term Symbol Ann, Type Symbol Ann))
-        hydrateTerm codebaseEnv ref =
-          Codebase.codebaseMToTransaction codebaseEnv do
-            term <- Codebase.expectTerm ref
-            pure (ref, term)
-        hydrateType ::
-          Codebase.CodebaseEnv ->
-          TypeReferenceId ->
-          PG.Transaction e (TypeReferenceId, Decl Symbol Ann)
-        hydrateType codebaseEnv ref =
-          Codebase.codebaseMToTransaction codebaseEnv do
-            type_ <- Codebase.expectTypeDeclaration ref
-            pure (ref, type_)
+    let hydrateTerms ::
+          UserId ->
+          BiMultimap Referent Name ->
+          PG.Transaction e (Map Name (TermReferenceId, (Term Symbol Ann, Type Symbol Ann)))
+        hydrateTerms codebaseUser termReferents = do
+          let termReferenceIds = Map.mapMaybe Referent.toTermReferenceId (BiMultimap.range termReferents)
+          termIds <-
+            PG.pFor termReferenceIds \refId ->
+              (refId,) <$> DefnsQ.pipelinedExpectTermId refId
+          v2Terms <-
+            PG.pFor termIds \(refId, termId) ->
+              (refId,) <$> DefnsQ.pipelinedExpectTermById codebaseUser refId termId
+          v1Terms <-
+            for v2Terms \(refId, (term, typ)) ->
+              (refId,) <$> Codebase.convertTerm2to1 (Reference.idToHash refId) term typ
+          pure v1Terms
+        hydrateTypes ::
+          UserId ->
+          BiMultimap TypeReference Name ->
+          PG.Transaction e (Map Name (TypeReferenceId, Decl Symbol Ann))
+        hydrateTypes codebaseUser typeReferences = do
+          let typeReferenceIds = Map.mapMaybe Reference.toId (BiMultimap.range typeReferences)
+          typeIds <-
+            PG.pFor typeReferenceIds \refId ->
```

Comment:
I'm not positive, but I bet we could do even better than pipelining here and collect all the typeIds in a single batch query rather than a pipelined for.